Enhanced <input type=color>: infrastructure for alpha and colorspace attributes

annevk opened the pull request:
#### 8fc75a523e00c6976854d51ce31d4885364ee229
<pre>
Enhanced &lt;input type=color&gt;: infrastructure for alpha and colorspace attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281213">https://bugs.webkit.org/show_bug.cgi?id=281213</a>
<a href="https://rdar.apple.com/137663314">rdar://137663314</a>

Reviewed by Sam Weinig.

Add the necessary infrastructure for
<a href="https://github.com/whatwg/html/pull/10456.">https://github.com/whatwg/html/pull/10456.</a> In particular:

- Support the alpha and colorspace attributes behind a preference.
- When the preference is enabled vastly expand the CSS color parsing
  and serialization capabilities of &lt;input type=color&gt;.

Future changes will adjust the UI accordingly.

Tests are upstreamed here:
<a href="https://github.com/web-platform-tests/wpt/pull/48630">https://github.com/web-platform-tests/wpt/pull/48630</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.tentative-expected.txt: Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.tentative.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.html.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window.js: Added.
(const.assert_colors.set_up_fuzzy_color_test):
(forEach):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/w3c-import.log:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::colorParsingParameters):
(WebCore::parseColorValue):
(WebCore::serializeColorValue):
(WebCore::ColorInputType::fallbackValue const):
(WebCore::ColorInputType::sanitizeValue const):
(WebCore::ColorInputType::valueAsColor const):
(WebCore::ColorInputType::typeMismatchFor const):
(WebCore::ColorInputType::didChooseColor):
(WebCore::ColorInputType::suggestedColors const):
(WebCore::ColorInputType::selectColor):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::alpha):
(WebCore::HTMLInputElement::colorSpace):
(WebCore::HTMLInputElement::setColorSpace):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:

Canonical link: <a href="https://commits.webkit.org/285322@main">https://commits.webkit.org/285322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6862c0d5ea29626d527822f27bf476bcf57f707a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72149 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76315 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23358 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23180 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56901 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15406 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75216 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62140 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43403 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19614 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21708 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65278 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65309 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77994 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71403 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16390 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19139 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65370 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16437 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62163 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64629 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15957 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12834 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6482 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93184 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47368 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2152 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20519 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48437 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49725 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48180 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -86,6 +93,56 @@ static std::optional<SRGBA<uint8_t>> parseSimpleColorValue(StringView string)
+            input = inputAsRGBA.resolved();
+    } else {
+        // FIXME: Ideally this would not be a lossy conversion and preserve colors such as color(display-p3 2 2 2).
+        input = input.toColorTypeLossy<DisplayP3<float>>().resolved();
```
@weinig I'd appreciate your input on this FIXME and more generally on the integration with the color code if you have the time. Thanks!

weinig replied:
The name is a bit misleading. The conversion is only lossy if the requested color type would cause it to be lossy. So it's more like, `toColorType*Potentially*Lossy()`.

In this case though, you should use `ExtendedDisplayP3` rather than `DisplayP3`, so there is no clamping applied.

--

I would probably put this in an else if (colorSpace == "display-p3") block and/or assert that is the color space, as its a bit tricky to locally reason about why using DisplayP3 here is right.

annevk commented on Source/WebCore/html/HTMLInputElement.idl:
```diff
@@ -25,10 +25,12 @@
     Exposed=Window
 ] interface HTMLInputElement : HTMLElement {
     [CEReactions=NotNeeded, Reflect] attribute DOMString accept;
+    [EnabledBySetting=InputTypeColorEnhancementsEnabled, CEReactions=NotNeeded, Reflect] attribute boolean alpha;
```
I need to add `Conditional=INPUT_TYPE_COLOR` here as the binding layer is not that smart.

weinig commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -86,6 +93,56 @@ static std::optional<SRGBA<uint8_t>> parseSimpleColorValue(StringView string)
+        if (alpha)
+            input = { inputAsRGBA, { Color::Flags::UseColorFunctionSerialization } };
+        else
+            input = inputAsRGBA.resolved();
```
I think a comment explaining these branches and the handling of alpha would be helpful. Its not immediately obvious by when there is alpha the color function serialization flag is needed.

weinig commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -124,21 +181,27 @@ bool ColorInputType::supportsRequired() const
     ASSERT(element());
-    return parseSimpleColorValue(element()->value()).value();
+    auto color = parseColorValue(element()->value(), *element());
+    ASSERT(!!color);
```
What makes this assert ok? It seems like parseColorValue() can return nullopt if the parse fails.

annevk replied:
This is pre-existing code I'm trying to clarify. `parseSimpleColorValue()` can also return nullopt. Given that value is set to a fallback value if parsing fails on setting this should always parse successfully.

weinig commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -86,6 +93,56 @@ static std::optional<SRGBA<uint8_t>> parseSimpleColorValue(StringView string)
+    if (context.colorSpace().isNull())
+        return parseSimpleColorValue(string);
+
+    auto color = CSSPropertyParserHelpers::parseColorRaw(string.toString(), CSSParserContext(HTMLStandardMode), [&] {
```
Since you have a Document available (`context.document()`), you can use a properly initialized `CSSParserContext` here.

```
auto parserContext = CSSParserContext(context.document());
```

(or something along those lines).

pxlcoder replied:
Since you have a Document, `context.document().cssParserContext()` should work.

weinig reviewed (approved):
Looks good. Should be testable though.

aprotyas commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -86,6 +93,59 @@ static std::optional<SRGBA<uint8_t>> parseSimpleColorValue(StringView string)
+
+    auto parserContext = context.document().cssParserContext();
+    parserContext.mode = HTMLStandardMode;
+    auto color = CSSPropertyParserHelpers::parseColorRaw(string.toString(), parserContext, [&] {
```
Nit: you don't need to capture anything in this lambda.

aprotyas reviewed (commented):
LGTM -- It seems like the iOS failures are legit?

annevk committed:
Enhanced <input type=color>: infrastructure for alpha and colorspace attributes
https://bugs.webkit.org/show_bug.cgi?id=281213
rdar://137663314

Reviewed by Sam Weinig.

Add the necessary infrastructure for
https://github.com/whatwg/html/pull/10456. In particular:

- Support the alpha and colorspace attributes behind a preference.
- When the preference is enabled vastly expand the CSS color parsing
  and serialization capabilities of <input type=color>.

Future changes will adjust the UI accordingly.

Tests are upstreamed here:
https://github.com/web-platform-tests/wpt/pull/48630

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-attributes.window.js: Added.
(test):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.tentative-expected.txt: Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.tentative.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.html.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/color.window.js: Added.
(const.assert_colors.set_up_fuzzy_color_test):
(forEach):
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/w3c-import.log:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/ColorInputType.cpp:
(WebCore::colorParsingParameters):
(WebCore::parseColorValue):
(WebCore::serializeColorValue):
(WebCore::ColorInputType::fallbackValue const):
(WebCore::ColorInputType::sanitizeValue const):
(WebCore::ColorInputType::valueAsColor const):
(WebCore::ColorInputType::typeMismatchFor const):
(WebCore::ColorInputType::didChooseColor):
(WebCore::ColorInputType::suggestedColors const):
(WebCore::ColorInputType::selectColor):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::attributeChanged):
(WebCore::HTMLInputElement::alpha):
(WebCore::HTMLInputElement::colorSpace):
(WebCore::HTMLInputElement::setColorSpace):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:

Canonical link: https://commits.webkit.org/285322@main
